packages/jupyter-ai/src/inline-completions.ts

```diff
@@ -0,0 +1,348 @@
+        break;
+      }
+      case 'connection': {
+        this._initialized.resolve();
+        break;
+      }
+      default: {
+        if (message.reply_to in this._replyForResolver) {
+          this._replyForResolver[message.reply_to](message);
+          delete this._replyForResolver[message.reply_to];
+        } else {
+          console.warn('Unhandled message', message);
+        }
+        break;
+      }
+    }
+  }
+
+  /**
+   * Dictionary mapping message IDs to Promise resolvers.
+   */
+  private _replyForResolver: Record<
+    number,
+    (value: AiService.InlineCompletionReply) => void
+  > = {};
+
+  private _onClose(e: CloseEvent, reject: any) {
+    reject(new Error('Inline completion websocket disconnected'));
+    console.error('Inline completion websocket disconnected');
+    // only attempt re-connect if there was an abnormal closure
+    // WebSocket status codes defined in RFC 6455: https://www.rfc-editor.org/rfc/rfc6455.html#section-7.4.1
+    if (e.code === 1006) {
+      const delaySeconds = 1;
+      console.info(`Will try to reconnect in ${delaySeconds} s.`);
+      setTimeout(async () => await this._initialize(), delaySeconds * 1000);
+    }
+  }
+
+  private async _initialize(): Promise<void> {
+    if (this.isDisposed) {
+      return;
+    }
+    const promise = new PromiseDelegate<void>();
+    this._initialized = promise;
+    console.log(
+      'Creating a new websocket connection for inline completions...'
+    );
+    const { token, WebSocket, wsUrl } = this.serverSettings;
+    const url =
+      URLExt.join(wsUrl, SERVICE_URL) +
+      (token ? `?token=${encodeURIComponent(token)}` : '');
+
+    const socket = (this._socket = new WebSocket(url));
+    socket.onclose = e => this._onClose(e, promise.reject);
+    socket.onerror = e => promise.reject(e);
+    socket.onmessage = msg => msg.data && this._onMessage(JSON.parse(msg.data));
+  }
+
+  private _isDisposed = false;
+  private _socket: WebSocket | null = null;
+  private _modelChanged = new Signal<CompletionWebsocketHandler, string>(this);
+  private _initialized: PromiseDelegate<void> = new PromiseDelegate<void>();
+}
+
+class JupyterAIInlineProvider implements IInlineCompletionProvider {
+  readonly identifier = 'jupyter-ai';
+
+  constructor(protected options: JupyterAIInlineProvider.IOptions) {
+    options.completionHandler.modelChanged.connect(
+      (_emitter, model: string) => {
+        this._currentModel = model;
+      }
+    );
+  }
+
+  get name() {
+    if (this._currentModel.length > 0) {
+      return `JupyterAI (${this._currentModel})`;
+    } else {
+      // This one is displayed in the settings.
+      return 'JupyterAI';
+    }
+  }
+
+  async fetch(
+    request: CompletionHandler.IRequest,
+    context: IInlineCompletionContext
+  ) {
+    const mime = request.mimeType ?? 'text/plain';
+    if (mime === 'text/x-ipythongfm') {
+      // Do not offer suggestions in markdown cells.
+      return { items: [] };
+    }
```

Comment:
We could support this, but a different template would be needed. It could be addressed in a future PR. Let me know what you think.